pkg/messaging/mqtt/pubsub_test.go

```diff
@@ -0,0 +1,599 @@
+// Copyright (c) Mainflux
+// SPDX-License-Identifier: Apache-2.0
+
+package mqtt_test
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/mainflux/mainflux/pkg/messaging"
+	mqtt_pubsub "github.com/mainflux/mainflux/pkg/messaging/mqtt"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	topic       = "topic"
+	chansPrefix = "channels"
+	channel     = "9b7b1b3f-b1b0-46a8-a717-b8213f9eda3b"
+	subtopic    = "engine"
+)
+
+var (
+	msgChan = make(chan messaging.Message)
+	data    = []byte("payload")
+)
+
+// Tests the MQTT Broker being used for the tests
+func TestMQTTBroker(t *testing.T) {
+	// Subscribing with topic, and with subtopic, so that we can publish messages
+	client, err := newClient(address, "clientID", 30*time.Second)
+	t.Cleanup(func() {
+		defer client.Unsubscribe()
+		defer client.Disconnect(5)
+	})
+	require.Nil(t, err, fmt.Sprintf("got unexpected error: %s", err))
+
+	token := client.Subscribe(topic, qos, mqttHandler(handler{false}))
+	token.Wait()
+	require.Nil(t, token.Error(), fmt.Sprintf("got unexpected error: %s", token.Error()))
+	token = client.Subscribe(fmt.Sprintf("%s.%s", topic, subtopic), qos, nil)
+	require.Nil(t, token.Error(), fmt.Sprintf("got unexpected error: %s", token.Error()))
+
+	cases := []struct {
+		desc    string
+		topic   string
+		message []byte
+	}{
+		{
+			desc:    "publish nil message to topic",
+			topic:   topic,
+			message: nil,
+		},
+		{
+			desc:    "publish nil message to topic.subtopic",
+			topic:   fmt.Sprintf("%s.%s", topic, subtopic),
+			message: nil,
+		},
+		{
+			desc:    "publishing to topic",
+			topic:   topic,
+			message: data,
+		},
+		{
+			desc:    "publishing to topic.subtopic",
+			topic:   fmt.Sprintf("%s.%s", topic, subtopic),
+			message: data,
+		},
+	}
+	for _, tc := range cases {
+		token = client.Publish(tc.topic, qos, false, tc.message)
+		token.Wait()
+		receivedMsg := <-msgChan
+		if tc.message == nil {
+			assert.Equal(t, 0, len(receivedMsg.Payload), fmt.Sprintf("%s: expected nil message, but got: %s", tc.desc, string(receivedMsg.Payload)))
+		} else {
+			assert.Equal(t, tc.message, receivedMsg.Payload, fmt.Sprintf("expected %s, got %s", tc.message, &receivedMsg.Payload))
+		}
+	}
+
+	token = client.Unsubscribe(topic)
+	token.Wait()
+	require.Nil(t, token.Error(), fmt.Sprintf("got unexpected error: %s", token.Error()))
+	token = client.Unsubscribe(fmt.Sprintf("%s.%s", topic, subtopic))
+	token.Wait()
+	require.Nil(t, token.Error(), fmt.Sprintf("got unexpected error: %s", token.Error()))
+}
+
+// Tests only the Publisher
+func TestPublisher(t *testing.T) {
```

Comment:
Add a case where you publish with empty topic

Reply:
@AryanGodara Do not forget this case.